docs/docs/guides/03-typescript.md

```diff
@@ -2,4 +2,226 @@
+
+- `import type` must be used for types
+- `import` must be used for the rest
+
+Example:
+
+```typescript
+// ❌ don't
+
+import { User, findUser } from './user';
+
+const user: User = findUser();
+
+// ✅ do
+
+import type { User } from './user';
+import { findUser } from './user';
+
+const user: User = findUser();
+```
+
+### c. Exports
+
+In most cases, it is recommended to use regular exports (export) instead of the default export (export default). Regular exports provide more flexibility and clarity when importing and using modules. They allow importing specific members of a module explicitly, which improves code readability and maintainability.
+
+However, there are a few exceptions where using export default might be appropriate. For example, when exporting a single primary entity or class from a module, or when working with modules that have a natural default export, such as a configuration object.
+
+Using regular exports provides a clearer and more explicit interface for module consumers, promoting better code organization and maintainability. However, it is important to consider the context before deciding whether to use default exports.
+
+## 3. Generic Types' Parameters
+
+When working with generic types, it is required to follow naming conventions and use descriptive names for type parameters. This helps improve code readability and maintainability over the whole codebase.
+
+- For simple types with a single self-explanatory parameter, it is allowed to **use `T`**
+- For any other generic types' parameters, it is required to **use a meaningful PascalCased variable name prefixed by `T`**
+
+Example:
+
+```typescript
+// ❌ don't
+function query<T extends object, U extends keyof T>(data: T, properties: U[]): { [K in U]: T[K] } {
+  // ...
+}
+
+// ✅ do
+function query<TValue extends object, TKey extends keyof TValue>(
+  data: TValue,
+  properties: TKey[]
+): { [TIter in TKey]: TValue[TIter] } {
+  // ...
+}
+
+// ❌ avoid
+type NonEmpty<TValue extends string> = TValue extends '' ? never : TValue;
+
+// ✅ prefer
+type NonEmpty<T extends string> = T extends '' ? never : T;
+```
+
+Choose meaningful names for type parameters that accurately describe their purpose within the context of the function or type.
+
+## 4. Avoiding Breaking Changes for Users
+
+When making changes to the Strapi types, it is important to consider the impact on existing users. To avoid breaking changes, follow these guidelines:
+
+- **Avoid removing or changing public APIs:** Public types APIs include functions, classes, or interfaces/types exposed to users as part of the official/documented Strapi API. Removing or changing them without providing appropriate deprecations or migration paths can break users' existing code.
+- **Use deprecation warnings instead:** If you need to deprecate an existing API, provide deprecation warnings and suggest alternative approaches. Give users a reasonable amount of time to adapt their code to the changes.
```

Comment:
Same here, how would a type definition deprecation look for someone making that change?

Reply:
36ca09280f3d8d21609b3c66b8c424ed3885d5ca